### PR TITLE
Resolve symbol names using bcc_elf_foreach_sym

### PR DESCRIPTION
**Pull request by jasonkeene**
I had a discussion w/ BCC here:

https://github.com/iovisor/bcc/pull/2441

Apparently, `bcc_resolve_symname` is not a general purpose symbol resolution function. It is tailored toward BCC's specific use cases. Primarily, this prevents it from resolving symbols for global variables. The fallback to objdump works but since @palmtenor suggested to use `bcc_elf_foreach_sym` I figure I'd submit a PR for that.

I did some manual testing and ran the test suite and everything seems to work fine.

### CODE REVIEW

File: src/bpftrace.cpp [hunk 1]
```diff
@@ -1741,13 +1741,32 @@ uint64_t BPFtrace::resolve_cgroupid(const std::string &path) const
...
+  struct bcc_symbol_option option = {
+    .use_debug_file = 0,
+    .check_debug_file_crc = 0,
+    .use_symbol_type = (1 << STT_OBJECT),
+  };
+
+  struct bcc_symbol sym;
+  sym.name = name.c_str();
+
+  int err = bcc_elf_foreach_sym(path.c_str(), sym_resolve_callback, &option, &sym);
```
Comment (danobi) [inline src/bpftrace.cpp]: From reading the bcc code, I think bcc_elf_foreach_sym will still return 0 even if we never successfully resolve a symbol name. In this case, we'd be returning junk data from `sym.offset`, right?

**Comment by jasonkeene** [comment]
Yea, you are absolutely right. I read through the code and tested it with a test program and it gives nonsense as expected.

The docs in the header file say 0 means success or stopped by callback:

https://github.com/iovisor/bcc/blob/01ee0b61a854e06cb62272ddfbb4eeaf16e9ad7c/src/cc/bcc_elf.h#L57

It would be nice if they differentiated between stopped by callback and not.

The example I was linked to in BCC doesn't even check the return value:

https://github.com/iovisor/bcc/blob/01ee0b61a854e06cb62272ddfbb4eeaf16e9ad7c/examples/cpp/pyperf/PyPerfUtil.cc#L147-L148

I suppose, as a work-around, I could set `sym.offset` to `0`, run `bcc_elf_foreach_sym`, and then check if `sym.offset` is still `0`, and throw if it is. What do you think?

**Comment by danobi** [comment]
> I suppose, as a work-around, I could set sym.offset to 0, run bcc_elf_foreach_sym, and then check if sym.offset is still 0, and throw if it is. What do you think?

Sounds good. 0x0 seems invalid enough.

One other thing -- could you make the callback `static` or a lambda? Better to not leave global symbols hanging around.

**Comment by jasonkeene** [comment]
Sure thing.

**Comment by jasonkeene** [comment]
I just noticed another callback here:

https://github.com/iovisor/bpftrace/blob/ec8b61aac194b5f2dae48c90271198df3af123ed/src/bpftrace.cpp#L1765-L1769

Are these the global symbols you don't want?:

```
# readelf -sW /usr/local/bin/bpftrace | grep sym_resolve_callback
  3295: 00000000001c9812    84 FUNC    GLOBAL DEFAULT   14 _ZN8bpftrace20sym_resolve_callbackEPKcmmPv
 13054: 00000000001c9812    84 FUNC    GLOBAL DEFAULT   14 _ZN8bpftrace20sym_resolve_callbackEPKcmmPv
# readelf -sW /usr/local/bin/bpftrace | grep add_symbol
 11913: 00000000001c9a2a   224 FUNC    GLOBAL DEFAULT   14 _ZN8bpftrace10add_symbolEPKcmmPv
  3383: 00000000001c9a2a   224 FUNC    GLOBAL DEFAULT   14 _ZN8bpftrace10add_symbolEPKcmmPv
```

Seems like they are mangled already. Maybe because of the namespace? I assume you wanted them to be static methods on `BPFtrace` class to get mangling. I'm not super experienced with C++ so I might have misunderstood what you meant.

**Comment by danobi** [comment]
The `static` keyword on a function means the function cannot be accessed outside the file it was defined in. This is nice because it lets the compiler aggressively inline. Mostly just a good coding practice.

**Comment by danobi** [comment]
> I just noticed another callback here:

Yeah that should probably be static, a lambda, or in an anonymous namespace as well.

**Comment by jasonkeene** [comment]
Ah, cool. I wasn't aware of that. Making them static gave me some compiler warnings:

```
/build/bpftrace/src/bpftrace.cpp:1780:12: warning: ‘int bpftrace::add_symbol(const char*, uint64_t, uint64_t, void*)’ defined but not used [-Wunused-function]
 static int add_symbol(const char *symname, uint64_t /*start*/, uint64_t /*size*/, void *payload) {
            ^~~~~~~~~~
/build/bpftrace/src/bpftrace.cpp:1744:12: warning: ‘int bpftrace::sym_resolve_callback(const char*, uint64_t, uint64_t, void*)’ defined but not used [-Wunused-function]
 static int sym_resolve_callback(const char *name, uint64_t addr, uint64_t, void *payload)
            ^~~~~~~~~~~~~~~~~~~~
```

I did some experimenting and it looks like the static analysis of the use of these functions evaluates both branches of the `#ifdef HAVE_BCC_ELF_FOREACH_SYM`. That is pretty cool! Probably why you wanted these to be static in the first place.

I decided to just `#ifdef` out both functions since they are not needed in the `#else` case.

**Pull request merged**